### DISCUSSION TITLE
Handling Paginated Requests

### DISCUSSION THREAD

**Discussion by Bullkn0x** [postmanlabs/postman-flows#217] [Q&A]
Hey Team,

Is there currently a way to deal with paginated requests at the moment in flows?  I am trying to collect the data on x pages by navigating to the corresponding "next" link in the "link" response header, and run a for each on the resulting data set.  I have been struggling for a couple days to find a viable solution and I'm curious if in it's current state it's possible?

Much appreciated team!

**Comment by saswatds** [comment]
Hey @Bullkn0x this is definitely a good use-case to support. There are usually two kinds of pagination strategies
* Numbers pages with offsets
* A link to the next page

Currently the number pages will be difficult to do. I will work more on this in the coming weeks to add arithmetic operators and maybe a dedicated Pagination block.

But if you have links in your response (I think this is your use-case), I was able to come up with something like this.

![image](https://user-images.githubusercontent.com/8796447/150393207-cee6bf3b-93b1-4d59-b95c-3cbc3cbcf3e1.png)
First Create Data block
![image](https://user-images.githubusercontent.com/8796447/150393307-e9b6c805-c8c6-442a-a0fa-6269fc12b057.png)
Second Create Data block
![image](https://user-images.githubusercontent.com/8796447/150393377-9be72b88-569d-4ff1-bfeb-3fb8398dfa83.png)

Within the request, you can make the url as variable `{{page}}`
Will this work for you?

**Reply by Bullkn0x** [reply]
Yeah this would work if the link was directly associated with a key in the response body or header,  but this is a Link header [see RFC-5988](https://datatracker.ietf.org/doc/html/rfc5988#section-5.5) such that the Value is a string that would needs to be parsed to identify it's components.  


GitHub uses these for pagination in their API, for example:

```python3
>>> url = 'https://api.github.com/users/kennethreitz/repos?page=1&per_page=10'
>>> r = requests.head(url=url)
>>> r.headers['link']
'<https://api.github.com/users/kennethreitz/repos?page=2&per_page=10>; rel="next", <https://api.github.com/users/kennethreitz/repos?page=6&per_page=10>; rel="last"'
```

Requests will automatically parse these link headers and make them easily consumable:
```python
>>> r.links["next"]
{'url': 'https://api.github.com/users/kennethreitz/repos?page=2&per_page=10', 'rel': 'next'}

>>> r.links["last"]
{'url': 'https://api.github.com/users/kennethreitz/repos?page=7&per_page=10',
```
I like the idea of a dedicated pagination block that would build a data structure containing the relational attributes and their corresponding links?

I don't think it's possible current state to achieve this.  I thought I may have had a janky solution by using test scripts to increment an environment variable that would be passed into page query param, but sadly test scripts and pre-request scripts don't run in flows, which I would argue would be a great feature to add to allow users to configure custom functionality alongside the block functions.  

Adding an arithmetic operator would be beneficial outside of the scope of pagination as well. I think the only Approach I have that may work is to create a data structure with a list of numbers [1,2,... maybe 60], use a for each block to iterate through each number and assign that to a variable being passed in as page until the response body is empty, at which point all records have been obtained? Might work for now, but I'm oping

**Reply by saswatds** [reply]
Thanks, @Bullkn0x. This is really helpful. I will get the team to work on the spec for supporting all kinds of pagination. This will push the no-code way of doing things even further. 

P.S. Like it's said, 'Anything that can be done with js will get done with js` but we really want to build a solution that truly adopts the visual way of programming. If allow javascript to be a thing in these early stages then all of these cool things would remain hidden behind js and not get an opportunity to get special domain treatment.  Please continue letting us know things that you had to implement in js, and we will work out a visual solution for that.
